### PR TITLE
IAM Lifecycle Events

### PR DESCRIPTION
**Pull request by nan-li**
Provide these events to users through `setIAMLifecycleHandler(OSIAMLifecycleHandler handler)`
- onWillDisplayInAppMessage (OSInAppMessage message)
- onDidDisplayInAppMessage (OSInAppMessage message)
- onWillDismissInAppMessage (OSInAppMessage message)
- onDidDismissInAppMessage (OSInAppMessage message)

Callouts:
- changed when impressions are sent (from start of animation to end of animation instead)
- known issue with launch URLs with `Dismiss on click` **unchecked**, causing IAM to to be dismissed without firing the lifecycle events for dismissing
- renamed OSInAppMessage to OSInAppMessageInternal
- OSInAppMessageInternal now inherits from OSInAppMessage (which contains only messageId)

---
This change is [Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405)

### CODE REVIEW

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSIAMLifecycleHandler.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.onesignal;
+
+public abstract class OSIAMLifecycleHandler {
+    public void onWillDisplayInAppMessage(OSInAppMessage message) {
+        OneSignal.Log(OneSignal.LOG_LEVEL.VERBOSE, "IAM Will Display");
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSIAMLifecycleHandler.java:32]: Migth be worth it to add the IAM id to the event log?

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 2]
```diff
@@ -344,6 +347,45 @@ void onMessageActionOccurredOnPreview(@NonNull final OSInAppMessage message, @No
         logInAppMessagePreviewActions(action);
     }
 
+    /*
+    ** IAM Lifecycle methods
+    ** The following methods call the public inAppMessageLifecycleHandler callbacks
+    */
+
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:354]: Can we keep the same style to other Javadoc comments? 

/*
 *
 */
 method()

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java [hunk 3]
```diff
@@ -344,6 +347,45 @@ void onMessageActionOccurredOnPreview(@NonNull final OSInAppMessage message, @No
...
+    void onMessageWillDisplay(@NonNull final OSInAppMessageInternal message) {
+        if (OneSignal.inAppMessageLifecycleHandler == null) {
+            logger.verbose("OSInAppMessageController onMessageWillDisplay: inAppMessageLifecycleHandler is null");
+            return;
+        }
+        OneSignal.inAppMessageLifecycleHandler.onWillDisplayInAppMessage(message);
```
Comment (Jeasmine) [inline OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:360]: We could avoid this static reference to OneSignal class. Can we set the handler directly in this controller? And avoid OneSignal responsibility of having this handler set

File: OneSignalSDK/unittest/src/test/java/com/onesignal/OneSignalPackagePrivateHelper.java [hunk 4]
```diff
@@ -541,29 +545,48 @@ public static String getShowingInAppMessageId() {
...
+
+   public static void onMessageWillDismiss(@NonNull final OSInAppMessageInternal message) {
+      OneSignal.getInAppMessageController().onMessageWillDismiss(message);
+   }
+
+   public static void onMessageDidDismiss(@NonNull final OSInAppMessageInternal message) {
+      OneSignal.getInAppMessageController().onMessageDidDismiss(message);
+   }
+
+
```
Comment (Jeasmine) [inline OneSignalSDK/unittest/src/test/java/com/onesignal/OneSignalPackagePrivateHelper.java:581]: clean extra \n 🙏

**Review by Jeasmine — COMMENTED** [review]
LGTM! small comments

File: OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java (hunk 3 above)
Comment (emawby) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:360]: Should we do the same thing with the IAM click handler while we are at it?

Comment (emawby) [reply OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java:360]: Also are we guaranteed to always have the same InAppMessageController for the entire life of the App? If not then we will still need to store the handler in OneSignal statically and inject it into the InAppMessageController upon initialization

**Review by nan-li — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405)* status: 0 of 17 files reviewed, 4 unresolved discussions (waiting on @emawby, @Jeasmine, @jkasten2, and @tanaynigam)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSIAMLifecycleHandler.java, line 32 at r7](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-MgNSgP989LyCDK1o_lU-r7-32:-MgkqAE_BzWqxORqHf0l:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/4155ce89c35f964c8a666c53ff12c097eddd7d72/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSIAMLifecycleHandler.java#L32)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

Migth be worth it to add the IAM id to the event log?
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java, line 354 at r7](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-MgNSgPAC93sNnFfJBR2-r7-354:-Mgkq9_V4MQxhhQd2udF:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/4155ce89c35f964c8a666c53ff12c097eddd7d72/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java#L354)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

Can we keep the same style to other Javadoc comments? 

/*
 *
 */
 method()


</blockquote></details>

Done.
___
*[OneSignalSDK/unittest/src/test/java/com/onesignal/OneSignalPackagePrivateHelper.java, line 581 at r7](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-MgNSgPAC93sNnFfJBR8-r7-581:-Mgkq8Sx08ubJ6IT3bJA:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/4155ce89c35f964c8a666c53ff12c097eddd7d72/OneSignalSDK/unittest/src/test/java/com/onesignal/OneSignalPackagePrivateHelper.java#L581)):*
<details><summary><i>Previously, Jeasmine (JNahui) wrote…</i></summary><blockquote>

clean extra \n 🙏 
</blockquote></details>

Done.

**Review by jkasten2 — CHANGES_REQUESTED** [review]
Left some comments for now. I am still reviewing so I may have some additional comments.

> Reviewed 1 of 17 files at r8, 1 of 15 files at r9, 1 of 3 files at r11, 1 of 8 files at r12, 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-:-Mh15RDFDRW5dSnuHP4k:b-6c2zsz)* status: 5 of 17 files reviewed, 12 unresolved discussions (waiting on @emawby, @Jeasmine, @nan-li, and @tanaynigam)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/InAppMessageView.java, line 524 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh1-2YyBcDJMPmou8dI:-Mh1-2YyBcDJMPmou8dJ:btz2o2g) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/InAppMessageView.java#L524)):*
> ```Java
>             @Override
>             public void onAnimationEnd(Animation animation) {
>                 // For Android 6 API 23 devices, waits until end of animation to set elevation of CardView class
> ```

nit, styling inconsistently - One `if` statement has curly braces while on does not.  
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java, line 744 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh0hD358KugNOA3EWfU:-Mh0hD358KugNOA3EWfV:b9qydfc) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java#L744)):*
> ```Java
>    }
> 
>    public static void setIAMLifecycleHandler(@Nullable OSIAMLifecycleHandler handler) {
> ```

@emawby @nan-li What do you think about spelling out IAM?, We never (or almost never) expose the abbreviation in the dashboard or documentation.
  
`setIAMLifecycleHandler` -> `seInAppMessageLifecycleHandler`
`OSIAMLifecycleHandler` -> `OSInAppMessageLifecycleHandler`
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java, line 40 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh10raECFnPkCmbL-OR:-Mh10raECFnPkCmbL-OS:b-epilfo) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java#L40)):*
> ```Java
>     protected String messageId;
> 
>     OSInAppMessage() {
> ```

We can remove this default constructor, as the one below passing in an id should always be used.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java, line 43 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh114MH7xps53r9CsCk:-Mh114MH7xps53r9CsCl:b-i9pn9j) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java#L43)):*
> ```Java
>     }
> 
>     OSInAppMessage(String messageId) {
> ```

Can we add `@NonNull`, a non-null id should always be provided.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java, line 47 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh11B-t8rDCY56pUDPA:-Mh11B-t8rDCY56pUDPB:b52auzf) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java#L47)):*
> ```Java
>     }
> 
>     public String getMessageId() {
> ```

Along with my comment above, let's add `@NonNull` if we can.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java, line 360 at r7](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-MgNSgPAC93sNnFfJBR2-r7-360:-Mh11gVE-DS6fyA00wbQ:be95c6f) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/4155ce89c35f964c8a666c53ff12c097eddd7d72/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java#L360)):*
<details><summary><i>Previously, emawby (Elliot Mawby) wrote…</i></summary><blockquote>

Also are we guaranteed to always have the same InAppMessageController for the entire life of the App? If not then we will still need to store the handler in OneSignal statically and inject it into the InAppMessageController upon initialization
</blockquote></details>

I believe we initialize `InAppMessageController` once and lives the whole lifetime of the app, we should double check to be sure. If so what we can do is have OneSigna.java call `InAppMessageController.setInAppMessageLifecycleHandler` to set it. As an app developer can set one at anytime, not just at construction time.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java, line 71 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh147nv9ofBz9CZyHC1:-Mh147nv9ofBz9CZyHC2:b-3tukhr) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java#L71)):*
> ```Java
> 
>     // dismissFired prevents onDidDismiss from getting called multiple times
>     // (was a known issue for Prompts)
> ```

We can remove "(was a known issue for Prompts)". This isn't important to how it works now.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java, line 74 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh14O7lCjyVkxJhYGXL:-Mh14O7lCjyVkxJhYGXM:b-1lvvq5) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java#L74)):*
> ```Java
>     private boolean dismissFired = false;
>     // closing prevents IAM being redisplayed when the activity changes during an actionHandler
>     // (customer reported issue)
> ```

Let's remove "(customer reported issue)"
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java, line 322 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh14lS133UsGwY6USOP:-Mh14lS133UsGwY6USOQ:b8i6kw) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java#L322)):*
> ```Java
>         if (lastActivityName == null)
>             showMessageView(null);
>         else if (!lastActivityName.equals(currentActivityName) && !closing) {
> ```

From a quick look it seems we should move the `!closing` check inside this `else if` block. Otherwise we are going to fall into the else below which I don't think needs to be done.

**Review by jkasten2 — COMMENTED** [review]
Finished reviewing, don't have any additional comments.

> Reviewed 5 of 15 files at r9, 6 of 8 files at r12, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-:-MhF-P1bBNaJkKvPYJ1g:bbi3t60)* status: all files reviewed, 12 unresolved discussions (waiting on @Jeasmine, @nan-li, and @tanaynigam)

**Review by nan-li — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405)* status: 4 of 17 files reviewed, 11 unresolved discussions (waiting on @emawby, @Jeasmine, @jkasten2, and @tanaynigam)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java, line 744 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh0hD358KugNOA3EWfU:-MhFrNSd1bX1GSfQCJJw:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignal.java#L744)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

@emawby @nan-li What do you think about spelling out IAM?, We never (or almost never) expose the abbreviation in the dashboard or documentation.
  
`setIAMLifecycleHandler` -> `seInAppMessageLifecycleHandler`
`OSIAMLifecycleHandler` -> `OSInAppMessageLifecycleHandler`
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java, line 40 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh10raECFnPkCmbL-OR:-MhFwFr55vG5FB-ZfcJR:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java#L40)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

We can remove this default constructor, as the one below passing in an id should always be used.
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java, line 43 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh114MH7xps53r9CsCk:-MhFx4VB1ZkzfYm4hUdN:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java#L43)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Can we add `@NonNull`, a non-null id should always be provided.
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java, line 47 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh11B-t8rDCY56pUDPA:-MhFx6gb0SySc10x6hmr:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessage.java#L47)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Along with my comment above, let's add `@NonNull` if we can.
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java, line 71 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh147nv9ofBz9CZyHC1:-MhFyzMM3xfdtZFKjJRC:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java#L71)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

We can remove "(was a known issue for Prompts)". This isn't important to how it works now.
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java, line 74 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh14O7lCjyVkxJhYGXL:-MhFyzmJ42q8KtWdUD31:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java#L74)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

Let's remove "(customer reported issue)"
</blockquote></details>

Done.
___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java, line 322 at r14](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-Mh14lS133UsGwY6USOP:-MhFyvK9Ean2xKnEkXk5:b-896fix) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/063a3924d8dbbcc15e247511c03030f5066e66dc/OneSignalSDK/onesignal/src/main/java/com/onesignal/WebViewManager.java#L322)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

From a quick look it seems we should move the `!closing` check inside this `else if` block. Otherwise we are going to fall into the else below which I don't think needs to be done.
</blockquote></details>

Done.

**Review by nan-li — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405)* status: 4 of 18 files reviewed, 11 unresolved discussions (waiting on @emawby, @Jeasmine, @jkasten2, and @tanaynigam)

___
*[OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java, line 360 at r7](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-MgNSgPAC93sNnFfJBR2-r7-360:-MhLEJq9DZSDNJ8WpzuJ:bdbrjn9) ([raw file](https://github.com/onesignal/onesignal-android-sdk/blob/4155ce89c35f964c8a666c53ff12c097eddd7d72/OneSignalSDK/onesignal/src/main/java/com/onesignal/OSInAppMessageController.java#L360)):*
<details><summary><i>Previously, jkasten2 (Josh Kasten) wrote…</i></summary><blockquote>

I believe we initialize `InAppMessageController` once and lives the whole lifetime of the app, we should double check to be sure. If so what we can do is have OneSigna.java call `InAppMessageController.setInAppMessageLifecycleHandler` to set it. As an app developer can set one at anytime, not just at construction time.
</blockquote></details>

OK I made these changes you mentioned.

**Review by jkasten2 — COMMENTED** [review]
> Reviewed 1 of 13 files at r27, 4 of 8 files at r30, 1 of 6 files at r34, 1 of 5 files at r38, 1 of 1 files at r39, 2 of 3 files at r40, 1 of 2 files at r41, 1 of 1 files at r42, 2 of 2 files at r43, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405)* status: all files reviewed, 5 unresolved discussions (waiting on @emawby, @Jeasmine, and @tanaynigam)

**Review by jkasten2 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-:-MhQvtV3BCtVViZwKW5v:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @Jeasmine and @tanaynigam)

**Review by tanaynigam — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/onesignal/onesignal-android-sdk/1405#-:-MhnxsxlAlWUrm5ktf8V:bnfp4nl)* status: all files reviewed, 4 unresolved discussions (waiting on @Jeasmine)

**Pull request merged**